### PR TITLE
refactor(web/engine): new RuleBehavior return type from keyboard calls, utilization

### PR DESCRIPTION
**Pull request by jahorton**
This PR is the first in a likely-to-be-long sequence of PRs aimed at our long-standing goal of enabling a headless KMW - the `web-core`, as it were.  (The exact form for the planned module's name is still pending.)  To make this happen, we'll need to refactor the `KeyboardInterface` and `Processor` classes in such a way that we can remove literally all references to the DOM.

One easy early target - KeyboardInterface has a few, though not many, references to the DOM - some of which plagued us due to their interaction with fat-fingering for predictive text.  This PR will focus on those parts.

Note that nearly all "changes" directly serve refactoring - all existing logic is preserved; the only question is _where_ that logic is evaluated.
- The one main logic changed regards how we determine whether or not to set `DOMEventHandlers.states.changed`, as we can now determine that from a `Transform`, easily outside of `KeyboardInterface` space.

### Reviewing Tip

I made sure to keep each commit very self-contained and step-by-step, save maybe for the TSS_STORE manipulation.  At any rate, examining each commit individually should be very helpful.

### On `LegacyKeyEvent`'s removal

A quick search through all of KeymanWeb shows that only the deleted references to that old type remain in KeymanWeb's code base.  Nothing generates the `LegacyKeyEvent`, so I removed it.  There's a chance this affects the CJK keyboards, though.

On investigation, it seems this was left in place previously as a "just in case" - all actual references _appear_ to have been resolved as early as 12.0, though it's hard to be absolutely 100% sure.  kmwembedded.ts used to supply... less "complete" KeyEvent structures.

### On `LeventMatched` -> "Rule Behavior"

In versions of KeymanWeb before 12.0, there was very little internal state-tracking management.  `Transcription`s and `Transforms` were then added in 12.0 to facilitate the basic state-tracking necessary to facilitate predictive text - if a rule was matched, we then examined the state of the DOM after the rule executed to determine the rule's effects.  At that point, we still relied on a simple `boolean` value to determine whether or not the rule triggered.

That stated... an object documenting the effects of a matched rule can itself fulfill that boolean flag.  If we don't match a rule, there are no effects; otherwise, we can instead have those effects determined somewhat more proactively.

With these changes in place, the `RuleBehavior` object tracks a few extra details compared to a raw `Transcription`:
- whether or not a "beep" feedback response should be generated by the rule.
- whether or not the rule itself (not the key!) triggered a layer change.

These effects should only be processed for keys that were actually pressed.  Now, _regardless of the `OutputTarget` type_, we can ensure that they always take effect for only the event-triggering key and not for any Alternates.
- Note: we may wish for these to be triggered by `Mock`s in our eventual target of an ideal "embedded" mode.  This change was necessary to support that.

Of course, having the rule actually produce these flags on a return object is also a big help for unit-testing.  That said, those unit tests will have to wait for a more complete 'headless'.

### Remaining References (KeyboardInterface)

- `doInputEvent`
    - This is used to force-trigger a DOM event for any listeners; worst-case, we can stub it out later and have the DOM-aware module extend the web-core, implementing it on web-core's behalf.
- `notifyKeyboard` has two references to `HTMLElement`.
    - This function is used by the CJK keyboards.
    - The only known uses are picker-based and the picker itself requires DOM access.
        - It may be a bit tricky, but 'extending' Processor appropriately by DOM-aware modules may suffice... once we can write the two modules separately.
- There also remain 4 references to our `DOMEventHandlers` class.
   - All of them are to ignore 'selection changes' necessary because of issues related to #539 and its planned resolution.

### User Testing

These changes have been tested against the following keyboards:
* `lao_2008_basic` - generates `beep` feedback
* `nrc_crk_cans` - has rules with layer manipulation output
    - Refer to https://github.com/keymanapp/keyman/issues/2162 for guidance.
* `chinese` - to ensure CJK keyboards still work.
* `japanese` - to ensure CJK keyboards still work.
  - The active input element _really_ hates losing focus, though that was also the case on `master`.

Also tested against `testing/prediction-ui`, ensuring that predictive text still works consistently as before.

All existing unit tests continue to pass.

### CODE REVIEW

**Comment by jahorton** [comment]
Just found a problem here - predictive suggestions aren't being properly applied when these changes are in place.

**Comment by jahorton** [comment]
OK, found the issue and fixed it; all is sane again.  Forgot to match the alternates to the same Transform id, since they have distinct ones by default from the `RuleBehavior` return.

File: web/source/kmwdom.ts [hunk 1]
```diff
@@ -82,6 +101,64 @@ namespace com.keyman {
...
+     * Function     beep          KB      (DOM-side implementation)
+     * Scope        Public
+     * @param       {Object}      Pelem     element to flash
+     * Description  Flash body as substitute for audible beep; notify embedded device to vibrate
+     */    
+    doBeep(outputTarget: text.OutputTarget) {
+      // Handles embedded-mode beeps.
+      let keyman = com.keyman.singleton;
+      if ('beepKeyboard' in keyman) {
+        keyman['beepKeyboard']();
```
Comment (mcdurdin) [inline web/source/kmwdom.ts]: If this is embedded mode only, surely we want to return after this?

File: web/source/kmwdom.ts [hunk 2]
```diff
@@ -82,6 +101,64 @@ namespace com.keyman {
...
+
+      for(var Lbo=0; Lbo<this._BeepObjects.length; Lbo++) { // I1446 - beep sometimes fails to return background color to normal
+                                                                  // I1511 - array prototype extended
+        if(this._BeepObjects[Lbo].e == Pelem) {
+          return;
+        }
+      }
+      
+      this._BeepObjects = com.keyman.singleton._push(this._BeepObjects, new BeepData(Pelem));
+      Pelem.style.backgroundColor = '#000000';
```
Comment (mcdurdin) [inline web/source/kmwdom.ts]: This is probably not the time for this, but of course if we are in 'dark mode', the beep will not be very visible!

Comment (jahorton) [reply web/source/kmwdom.ts]: I've gone ahead and made an issue for it.  I set it to "future" for now, since it seems like a low-priority item.

**Pull request merged**